### ISSUE TITLE
[PWA] Make use of the Window Controls Overlay feature to display menus in the title bar area

### ISSUE THREAD

**Issue by captainbrosset** [microsoft/vscode#140694]
The Window Controls Overlay PWA feature is coming out of origin trial and is shipping with Chrome/Edge 98.

To learn more about it, here are a few resources:

* https://alistapart.com/article/breaking-out-of-the-box/
* https://developer.mozilla.org/en-US/docs/Web/API/Window_Controls_Overlay_API
* https://docs.microsoft.com/en-us/microsoft-edge/progressive-web-apps-chromium/how-to/window-controls-overlay
* https://web.dev/window-controls-overlay/

In short, it allows PWAs installed on desktop OSes to regain control of the area normally reserved for the standard title bar. With it, the maximize/minimize/close/pwa buttons appear as overlays on top of the web content, and the rest of the title bar is gone.

It comes with a JS API and some CSS variables that help PWA make efficient use of the regained space.

Using this, the VS Code PWA could display its title bar at the top of the window, just like in its native version.
The hamburger menu version of it could still be used when the title bar area is too narrow (which is easy to detect with the ongeometrychange event).

**Comment by captainbrosset** [comment]
Pinging @digitarald who, I think, is looking at the PWA features right now.

**Comment by bpasero** [comment]
Fyi @sbatten

**Comment by digitarald** [comment]
Very much aligned with @sbatten's work of adding UI to the titlebar; and eventually optimizing that work for PWA.

**Comment by Offroaders123** [comment]
Any new information about this? WCO support would be amazing!

**Comment by bpasero** [comment]
Related: https://github.com/microsoft/vscode/issues/127449

**Comment by yume-chan** [comment]
![image](https://user-images.githubusercontent.com/1330321/171552547-bfa47b3b-f5ab-48cc-bc94-0229e9034f86.png)

I did a quick test. With only a little code change it already looks and works great. It really gives the feeling of native applications, if not seeing the two extra buttons at top-right added by the browser.

One remaining issue is that the title bar height is hardcoded now, it needs to be retrieved at runtime but I don't know how.

**Comment by bpasero** [comment]
Wow nice 👏

**Comment by captainbrosset** [comment]
> One remaining issue is that the title bar height is hardcoded now, it needs to be retrieved at runtime but I don't know how.

You can use the new CSS environment variables for this. In particular `env(titlebar-area-height)`. I wrote some docs about this here:  https://developer.mozilla.org/en-US/docs/Web/API/Window_Controls_Overlay_API#css_environment_variables

In fact, if you use these variables, you'll be able to position your menubar and other elements correctly across operating systems (including on Mac where the close/min/max buttons are on the left side).

> It really gives the feeling of native applications, if ignoring the two extra buttons at top-right added by the browser.

Yeah, these buttons are annoying when you want to achieve a "native" look. There's work going on right now about this. @diekus will surely be interested in this feedback as I know he's thinking about ways to get rid of the "chevron" icon.

**Comment by sbatten** [comment]
WCO also needs to be enabled in the browser. At least in edge, they are disabled by default.

**Comment by diekus** [comment]
It's coming very soon. Hope to have some news to share in some days! Once we have a timeframe, I am more than happy to help in any way to bring WCO to the the code PWA.

**Comment by bpasero** [comment]
@yume-chan curious, what was the change you did in https://github.com/microsoft/vscode/issues/140694#issuecomment-1144427460 to make it work?

**Comment by yume-chan** [comment]
My branch is at https://github.com/yume-chan/vscode/tree/pwa-wco (it's not conflicting with main).

**Comment by joaomoreno** [comment]
This is now enabled on https://insiders.vscode.dev. In Edge, make sure you enable the **Desktop PWA Window Controls Overlay** flag by visiting `edge://flags/#enable-desktop-pwas-window-controls-overlay`.

![image](https://user-images.githubusercontent.com/22350/200293001-e587c152-e8da-44fd-bf49-716713a96b9c.png)

@sbatten Should this look a little bit better?

![image](https://user-images.githubusercontent.com/22350/200293265-b919fa88-f3cf-434f-967e-284076d24dbe.png)

**Comment by diekus** [comment]
That's incredible news! Window Controls Overlay is now on by default in Chromium 105, so no flag needed! Congratulations @joaomoreno and team!

**Comment by sbatten** [comment]
@joaomoreno, yep, I thought I resolved this but I will force the title bar on and move the compact menu when in PWA mode as mentioned [here](https://github.com/microsoft/vscode/pull/165076)

**Comment by joaomoreno** [comment]
Looking good:

![image](https://user-images.githubusercontent.com/22350/201055400-9bcd6e10-0d8e-4440-8fb0-518fda3fa250.png)

**Comment by captainbrosset** [comment]
This is great to see! Amazing progress.

Just one detail:
I unfortunately forgot to grab a screenshot before dismissing it, but there's a notification banner that appears at the top of the window when you first install the app (it contains link to download VSCode, read the terms, etc.). This banner can't be closed after opting-in to WCO. Its close button is likely hidden behind the control buttons overlay. 
You might want to position it using the `env(titlebar-area-*)` variables.

**Comment by eifr** [comment]
Beautiful! small issue in mac but x100 better
![image](https://user-images.githubusercontent.com/5452613/201102856-3a83996d-05dc-472e-b30e-f65d90e7f221.png)

**Comment by sbatten** [comment]
@captainbrosset i have a fix going in today that should move the banner below the titlebar (like it is in desktop) when the WCO are shown.

**Comment by eifr** [comment]
Seems to not work on linux (fedora 36)

![Screenshot from 2022-11-13 12-33-57](https://user-images.githubusercontent.com/5452613/201517307-6f5eae2e-5f7b-46a0-bcd6-a024680b8a49.png)

**Comment by sbatten** [comment]
While this is generally done, macos web still has issues, so moving to December to resolve

**Comment by fogoplayer** [comment]
Loving the work on this feature. I'm running into some issues on the latest Insider Preview

Running on Chrome/ChromeOS 107.0.5304.110
[Screen recording 2022-12-01 12.21.24.webm](https://user-images.githubusercontent.com/13542612/205141376-2878e2f1-945e-479f-854e-696e910f3057.webm)

**Comment by leifwalsh** [comment]
What happened to this issue? It looked close to shipping but never did?

**Comment by fogoplayer** [comment]
It's live for me! On Chrome, look for an "up" Chevron near the window controls

**Comment by leifwalsh** [comment]
I don't see anything: 
![image](https://github.com/microsoft/vscode/assets/46406/9f5951d2-4b97-4a90-9c7e-5584a5d5f6bd)

This is on Windows 10 with Chrome 122.0.6261.95. Is it possible I've got something configured incorrectly?

**Comment by leifwalsh** [comment]
Oh, I do see it in Edge on Windows 11. Must be a local configuration difference!

**Comment by inLnx** [comment]
you have to zoom in to make it look even

**Comment by bpasero** [comment]
@leifwalsh this is a very good catch, looks like we are missing to set `"display_override": ["window-controls-overlay"]` in that case. A fix will be pushed today and should be available in insiders tomorrow.

**Comment by leifwalsh** [comment]
Awesome, thank you! Looking forward to the extra couple lines of editor space! :)

**Comment by bpasero** [comment]
@leifwalsh with todays insider, this landed. Can you maybe verify?

**Comment by leifwalsh** [comment]
yep, we see it, thank you!
